### PR TITLE
Only overflow the cache for one required frame

### PR DESCRIPTION
**Pull request by mklim**
Previously codec.cc needed all required frames to already be decoded
before it could decode any of their dependent frames. To accomplish this
it would always cache required frames, regardless of cache limit.

However both GIF and WEBP (the only currently supported animated image
formats) only allow the image to depend on one decoded frame at a time.
This means that there's no reason to cache all the required frames since
it's only valid for the image formats to require one previously decoded
frame at a time. (For example, frame 10 and frame 11 in a hypothetical
animated image could all depend on frame 9. But no subsequent frame
after frame 9 could depend on frames 0-8.)

Warning: this logic will break if we decide to support more animated
formats in the future.

Fixes flutter/flutter#24835.

### CODE REVIEW

**Comment by mklim** [comment]
:warning: This is ready for review but shouldn't be merged yet. :warning:

File: lib/ui/painting/codec.cc [hunk 1]
```diff
@@ -363,6 +363,43 @@ void Codec::dispose() {
...
+  timesUsed_ = 0;
+  const size_t cachedFrameSize = bitmap->computeByteSize();
+  // Don't cache this if we already have an entry or if it's non required and
+  // would overflow the cache.
+  if (bitmap_ ||
+      (numDependentFrames_ == 0 && codec_->OverflowsCache(cachedFrameSize))) {
+    return;
+  }
+  bitmap_ = std::make_unique<SkBitmap>(*bitmap);
+  codec_->decodedCacheSize_ += cachedFrameSize;
```
Comment (amirh) [inline lib/ui/painting/codec.cc]: If this overflows the cache and we have cached a non-required frame would be nice to evict the non required frame.

(Probably out of scope for this PR, if you think it makes sense a TODO will make it)

File: lib/ui/painting/codec.cc [hunk 2]
```diff
@@ -376,12 +413,14 @@ MultiFrameCodec::MultiFrameCodec(std::unique_ptr<SkCodec> codec,
   for (size_t frameIndex = 0; frameIndex < frameInfos_.size(); frameIndex++) {
     const auto& frameInfo = frameInfos_[frameIndex];
     if (frameInfo.fRequiredFrame != SkCodec::kNoFrame) {
-      frameBitmaps_[frameInfo.fRequiredFrame] =
-          std::make_unique<DecodedFrame>(/*required=*/true);
+      if (frameBitmaps_[frameInfo.fRequiredFrame] == nullptr) {
```
Comment (amirh) [inline lib/ui/painting/codec.cc]: Wouldn't we have always already set the frameBitmaps_ entry for the required frame (in line 423)

File: lib/ui/painting/codec.cc [hunk 3]
```diff
@@ -397,49 +436,58 @@ int MultiFrameCodec::repetitionCount() {
...
+
+  SkCodec::Options options;
+  options.fFrameIndex = frame;
+  const int requiredFrame = frameInfos_[frame].fRequiredFrame;
+  if (requiredFrame != SkCodec::kNoFrame) {
+    DecodedFrame& requiredCacheEntry = *frameBitmaps_[requiredFrame];
+    SkBitmap requiredBitmap;
+    if (!requiredCacheEntry.CachedBitmapExists()) {
+      FML_LOG(ERROR) << "Cache miss on required frame " << requiredFrame;
+      requiredBitmap = SkBitmap();
```
Comment (amirh) [inline lib/ui/painting/codec.cc]: Wouldn't default constructor already be invoked when you create the stack variable?

File: lib/ui/painting/codec.cc [hunk 4]
```diff
@@ -397,49 +436,58 @@ int MultiFrameCodec::repetitionCount() {
...
+      DecodePixels(&requiredBitmap, requiredFrame);
+    } else {
+      requiredBitmap = *requiredCacheEntry.ConsumeCachedBitmap();
```
Comment (amirh) [inline lib/ui/painting/codec.cc]: Maybe just make requiredBitmap a pointer so we only construct an empty one when needed?

File: lib/ui/painting/codec.cc [hunk 5]
```diff
@@ -397,49 +436,58 @@ int MultiFrameCodec::repetitionCount() {
...
+  const SkImageInfo info = codec_->getInfo().makeColorType(kN32_SkColorType);
+  bitmap->allocPixels(info);
+
+  SkCodec::Options options;
+  options.fFrameIndex = frame;
+  const int requiredFrame = frameInfos_[frame].fRequiredFrame;
+  if (requiredFrame != SkCodec::kNoFrame) {
+    DecodedFrame& requiredCacheEntry = *frameBitmaps_[requiredFrame];
+    SkBitmap requiredBitmap;
+    if (!requiredCacheEntry.CachedBitmapExists()) {
```
Comment (amirh) [inline lib/ui/painting/codec.cc]: Shouldn't the required frame always be cached at this point? (we are always caching required frames, and only evicting them after their last usage AFAIU)

File: lib/ui/painting/codec.h [hunk 6]
```diff
@@ -65,14 +67,22 @@ class MultiFrameCodec : public Codec {
...
-    DecodedFrame(bool required);
-    ~DecodedFrame();
+  // Handles logic based around caching required frames for future dependent
+  // frames.
+  class DecodedFrame {
+   public:
+    DecodedFrame(MultiFrameCodec* parent) : codec_(parent){};
+    const SkBitmap* ConsumeCachedBitmap();
+    bool CachedBitmapExists();
+    void MaybeCache(SkBitmap* bitmap);
```
Comment (amirh) [inline lib/ui/painting/codec.h]: nit: I found "maybe" here a bit confusing before reading the implementation, though I cannot think of a good name ("CacheIfRequiredOrHasSpace"  is my best attempt 😄 )

I think a comment here would help.

File: lib/ui/painting/codec.h [hunk 7]
```diff
@@ -65,14 +67,22 @@ class MultiFrameCodec : public Codec {
...
-  // other dependent frames.
-  struct DecodedFrame {
-    std::unique_ptr<SkBitmap> bitmap_ = nullptr;
-    const bool required_;
-
-    DecodedFrame(bool required);
-    ~DecodedFrame();
+  // Handles logic based around caching required frames for future dependent
+  // frames.
+  class DecodedFrame {
```
Comment (amirh) [inline lib/ui/painting/codec.h]: I'd suggest documenting the caching logic here as it isn't trivial.

File: lib/ui/painting/codec.h [hunk 8]
```diff
@@ -65,14 +67,22 @@ class MultiFrameCodec : public Codec {
...
+   public:
+    DecodedFrame(MultiFrameCodec* parent) : codec_(parent){};
+    const SkBitmap* ConsumeCachedBitmap();
+    bool CachedBitmapExists();
+    void MaybeCache(SkBitmap* bitmap);
+    void MaybeClearCache();
```
Comment (amirh) [inline lib/ui/painting/codec.h]: nit: Same comment here about "Maybe", what do you think about calling this something like "ConsumeRequiredFrame" ? (I'd rename "ConsumeCachedBitmap" to "GetCachedBitmap" as it isn't "consuming" the frame from the cache)

File: lib/ui/painting/codec.cc (hunk 1 above)
Comment (mklim) [reply lib/ui/painting/codec.cc]: Added a TODO at the top of this function and in the header for `DecodedFrame`. We'll probably want to follow up on this, should just be aware that at this point we'll start impacting performance and CPU usage in some cases to keep memory usage down.

File: lib/ui/painting/codec.cc (hunk 2 above)
Comment (mklim) [reply lib/ui/painting/codec.cc]: I think so, but wasn't totally sure. I was putting this check here in case for some reason there's a looping image and frame 0 depends on the very last frame. It seems unlikely but I couldn't find anything to rule it out. Do you know if they're always sequential or not?

**Comment by mklim** [comment]
Turns out the required frame dependencies are much more limited than what I originally thought and this PR can be hugely simplified, thanks @amirh. 🎉 Updated with a new patch to only cache one required frame over the limit at a time.

This is also now ready to merge whenever it gets LGTM.

**Comment by amirh** [comment]
Thanks Michael! great to see this simplified!

What do you think about further simplifying by removing our initialization step of marking required frames?
I believe we can instead just keep 2 members (one for the previous required frame and one for the previous frame) which should be enough to make sure we always have the required frame in memory.
We should then be able to treat all other frames with equal caching priority.

**Comment by mklim** [comment]
> I believe we can instead just keep 2 members (one for the previous required frame and one for the previous frame) which should be enough to make sure we always have the required frame in memory.

Just to make sure @amirh: you mean these two members in addition to the regular map of cached bitmaps within our decoded limit, right? Since we'd still want the frame-indexed cache for non-required frames after looping.

If I'm understanding your suggestion right, I think this would work but it would end up taking more memory at runtime. I think we'd need to cache 2 frames over the limit instead of just 1 in cases where we were using required frames at all, and in cases where we weren't we'd always cache 1 frame over the limit instead of 0. We'd want to always cache the N-1 frame in case it ends up being required, even in GIFs that don't have any required frames at all or only one required frame. Since we'd no longer be saving this info about the frames at initialization, we'd need to assume the worst. In cases of overflow always caching N-1 would also mean inserting and deleting from the cache every frame, too. And then on top of that we'd always want to cache the actual last required frame (if it exists) in case a previously required frame is used again instead of the N-1 frame. I don't think it would be preferable to the initialization step since it would end up using more memory, and I don't think the initialization step is really costing us much. What do you think?

**Comment by mklim** [comment]
We talked about this offline. So far the `frameBitmaps_` cache has been mixing concerns about caching extraneous frames under the limit and caching required frames in general, which makes the caching logic hard to follow because it's special casing the required frames. I didn't remove the initialization step entirely, but I did update the patch so that it always keeps the last required frame completely separate from the `frameBitmaps_` collection of generic decoded frames. I think the logic is more straightforward now.

File: lib/ui/painting/codec.cc [hunk 9]
```diff
@@ -371,20 +371,16 @@ MultiFrameCodec::MultiFrameCodec(std::unique_ptr<SkCodec> codec,
   compressedSizeBytes_ = codec_->getInfo().computeMinByteSize();
```
[the remark this answers is not in the document]
Comment (mklim) [reply lib/ui/painting/codec.cc]: clang-format is controlling the indentation here. We can set it off for certain lines if we want to tweak it though. I'm going to leave this as-is for now since it's not related to this PR, can alter this in a followup.
